Src/Plugins/Security/SEC101_043.NuGetCredentialsValidator.cs

```diff
@@ -0,0 +1,154 @@
+
+            XmlDocument hostXml = new XmlDocument();
+            hostXml.LoadXml(hostXmlAsString);
+
+            // hostXml looks like "<packageSources>...</packageSources>
+            List<string> hosts = hostXml
+                                    ?.ChildNodes[0] // <packageSources>...
+                                        ?.ChildNodes.Cast<XmlNode>().Where(x => x.Name.Equals("add", System.StringComparison.OrdinalIgnoreCase)) // <add ... >  <clear/> (the first node)
+                                            .Select(x => x.Attributes["value"]?.Value ?? x.Attributes["Value"]?.Value).ToList(); // <add key="name of host" value="http://nugetfeedUrl.com" /> (we're looking for URL)
+
+            // Uusually there will probably only be a single host.
+            // In any case, attempt to pass the credentials to each host one by one.
+            using (HttpClient httpClientNoCredentials = CreateHttpClient())
+            using (HttpClient httpClientWithCredentials = CreateHttpClient())
+            {
+                byte[] byteArray = Encoding.ASCII.GetBytes($"{username}:{password}");
+                httpClientWithCredentials.DefaultRequestHeaders.Authorization =
+                    new AuthenticationHeaderValue("Basic", Convert.ToBase64String(byteArray));
+
+                foreach (string host in hosts)
+                {
+                    if (!Uri.IsWellFormedUriString(host, UriKind.Absolute))
+                    {
+                        continue;
+                    }
+
+                    using (HttpResponseMessage responseWithNoCredentials = httpClientNoCredentials
+                    .GetAsync(host, HttpCompletionOption.ResponseHeadersRead)
+                    .GetAwaiter()
+                    .GetResult())
+                    {
+                        switch (responseWithNoCredentials.StatusCode)
+                        {
+                            case HttpStatusCode.OK:
+                                // Credentials not needed, this method of verification is indeterminate.
+                                return ReturnUnknownAuthorization(ref message, username);
+                            case HttpStatusCode.Unauthorized:
+                            case HttpStatusCode.Forbidden:
+                                // Credentials may resolve this, try again with them.
+                                using (HttpResponseMessage responseWithCredentials = httpClientWithCredentials
+                                .GetAsync(host, HttpCompletionOption.ResponseHeadersRead)
+                                .GetAwaiter()
+                                .GetResult())
+                                {
+                                    switch (responseWithCredentials.StatusCode)
+                                    {
+                                        case HttpStatusCode.OK:
+                                            // Credentials resolved the forbidden/unauthorized message we received.
```

Comment:
>// Credentials resolved the forbidden/unauthorized message we received. [](start = 44, length = 71)

This isn't a particularly helpful comment. You should avoid comments that are a sort of english language description of what is obvious in the code itself. It is our goal for methods like 'ReturnAuthorizedAccess' to be sufficiently descriptive to make a comment like 'The credentials are valid, return Authorized', and similar.

My test for a code comment is whether the comment has given me new information, which is both accurate and useful.

Reply:
I guess it's hard to tell what's obvious in the code.  Eddy mentioned that he didn't care for the nested switch statements so I wanted to be pretty verbose if I was going to contnue to use them.  Plus, even if it's likely obvious, I do think there is value in re-affirming that the reader's thought process is still aligned with the author's intent.  I also use nesting depth as a measure of complexity, and we're half way across the page at this point hahaha.  Don't you think some paranoia regarding clarity is warranted?

---
In reply to: [611240191](https://github.com/microsoft/sarif-pattern-matcher/pull/372#discussion_r611240191) [](ancestors = 611240191)